src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -183,6 +122,153 @@ public GeneratorDriver AddAdditionalTexts(ImmutableArray<AdditionalText> additio
+            {
+                var generator = _state.Generators[i];
+                var generatorState = _state.GeneratorStates[i];
+
+                results.Add(new GeneratorRunResult(generator, getSourceResults(generatorState), generatorState.Diagnostics, generatorState.Exception));
+                trees.AddRange(generatorState.Trees);
+                diagnostics.AddRange(generatorState.Diagnostics);
+            }
+
+            return new GeneratorDriverRunResult(results.ToImmutableAndFree(), trees.ToImmutableAndFree(), diagnostics.ToImmutableAndFree());
+
+            static ImmutableArray<GeneratedSourceResult> getSourceResults(GeneratorState generatorState)
+            {
+                ArrayBuilder<GeneratedSourceResult> sourceResults = ArrayBuilder<GeneratedSourceResult>.GetInstance();
+                for (int i = 0; i < generatorState.SourceTexts.Length; i++)
+                {
+                    var sourceText = generatorState.SourceTexts[i];
+                    var tree = generatorState.Trees[i];
+
+                    sourceResults.Add(new GeneratedSourceResult(tree, sourceText.Text, sourceText.HintName));
+                }
+                return sourceResults.ToImmutableAndFree();
+            }
+        }
+
+        internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, DiagnosticBag diagnosticsBag, CancellationToken cancellationToken = default)
+        {
+            // with no generators, there is no work to do
+            if (_state.Generators.IsEmpty)
+            {
+                return _state;
+            }
+
+            // run the actual generation
+            var state = StateWithPendingEditsApplied(_state);
+            var stateBuilder = ArrayBuilder<GeneratorState>.GetInstance();
+            int receiverCount = 0;
+
+            for (int i = 0; i < state.Generators.Length; i++)
+            {
+                var generator = state.Generators[i];
+                var generatorState = state.GeneratorStates[i];
+
+                // initialize the generator if needed
+                if (!generatorState.Info.Initialized)
+                {
+                    InitializationContext context = new InitializationContext(cancellationToken);
+                    Exception? ex = null;
+                    try
+                    {
+                        generator.Initialize(context);
+                    }
+                    catch (Exception e)
+                    {
+                        ex = e;
+                    }
+                    generatorState = ex is null
+                                     ? new GeneratorState(context.InfoBuilder.ToImmutable())
+                                     : SetGeneratorException(new GeneratorState(), generator, ex, diagnosticsBag, isInit: true);
+                }
+
+                // create the syntax receiver if requested
+                if (generatorState.Info.SyntaxReceiverCreator is object)
```

Comment:
Is `null` if initialize threw an exception? Consider asserting.

Reply:
I'm not sure I see the value. It's either null because it threw or because there wasn't one. either way we only care if there is one.

---
In reply to: [466072853](https://github.com/dotnet/roslyn/pull/46296#discussion_r466072853) [](ancestors = 466072853)